File: src/main/java/de/cplaiz/activecraftcore/listener/GamemodeChangeListener.java
package de.cplaiz.activecraftcore.listener;

import de.cplaiz.activecraftcore.playermanagement.Profilev2;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerGameModeChangeEvent;

public class GamemodeChangeListener implements Listener {

    @EventHandler
    public void onGamemodeChange(PlayerGameModeChangeEvent event) {
        Player player = event.getPlayer();
        Profilev2 profile = Profilev2.of(player);
        if (!profile.isFly()) return;
        boolean wasFlying = player.isFlying();
        Thread thread = new Thread(() -> {
            while (true)
                if (player.getGameMode() == event.getNewGameMode()) break;
            player.setAllowFlight(true);
            player.setFlying(wasFlying);
        });
        thread.start();
    }
}

File: src/main/java/org/activecraft/activecraftcore/listener/GamemodeChangeListener.kt
package org.activecraft.activecraftcore.listener

import org.activecraft.activecraftcore.playermanagement.Profilev2.Companion.of
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerGameModeChangeEvent

class GamemodeChangeListener : Listener {
    @EventHandler
    fun onGamemodeChange(event: PlayerGameModeChangeEvent) {
        val player = event.player
        val profile = of(player)
        if (!profile.isFly) return
        val wasFlying = player.isFlying
        val thread = Thread {
            while (true) if (player.gameMode == event.newGameMode) break
            player.allowFlight = true
            player.isFlying = wasFlying
        }
        thread.start()
    }
}